### ISSUE TITLE
Why `Tapeable*`?

### ISSUE THREAD

**Issue by bwdeng20** [LoadingByte/are-gnn-defenses-robust#3]
Thanks for sharing the source code! Could you please talk about something about why  `Parameter`, `Modules`, and `Optimizers` instead of native pytorch stuffs are used here?

**Comment by LoadingByte** [comment]
In brief, the standard module and optimizer classes are not suited to preserve gradient information about how parameter updates came to be. However, this information is needed to differentiate through the training procedure (which is done by Metattack and related attacks).

To solve that issue, we replace the standard PyTorch classes with out own versions. Notable changes are:

- Tensors wrapped by our `TapeableParameter` class may have a gradient.
- Our tapeable optimizers avoid inplace operations and the `no_grad()` context.

Does this help?
